### PR TITLE
Add HMR and split webpack configuration

### PR DESCRIPTION
**Pull request by ismaeldcom**
#### Changes proposed in this Pull Request
- Update `@wordpress/dependency-extraction-webpack-plugin` to support react refresh extraction.
- Installed the following required dependencies:
  - `@pmmmwh/react-refresh-webpack-plugin@0.5.7` – To load the runtime on our builds.
  - `react-refres@0.14.0` - The actual React Refresh code, used for Babel.
  - `webpack-dev-server@4.9.3` – Required to serve HMR changes.
  - `webpack-merge@5.8.0` – Used to easily merge different configurations effectively overriding some rules.
- Split webpack configuration into `shared`, `production`, and `development`– Adding HMR requires a bunch of settings that will clutter our production one, having them separated will make updates easier.
- Add HMR webpack configuration conditionally based on `WEBPACL_SERVE` from `env`.
- Load runtime script for admin area. As single runtime is required to make HMR work while using multiple entries on the same page. It will only load if `SCRIPT_DEBUG` is enabled and the runtime script exists.
- Add `npm run hmr` command with instructions.

I'd like to highlight the need for the runtime script and how it's loaded from the PHP side. Ideally we could get rid of it joining the runtime and a vendor chunk together, but for that we will need to properly configure split chunks and try to slim down or bundled dependencies. It's something more suited for a bundle optimization PR, where we could greatly reduce our bundle size.

#### Demo
https://user-images.githubusercontent.com/7670276/182116957-7e027bec-cc9d-456c-8d3d-98633bc45c99.mov

#### Testing instructions
The current implementation relies on WP loaded React Refresh, so it requires at least WordPress 6.

- Run `npm run build:release` before and after applying this branch.
- Compare the output, apart from the hashes everything else should match.
- Run `npm run hmr`.
- Navigate to a React page like WCPay settings or Multi-Currency.
- Update something from a mounted component.
- It should update without refreshing the page.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : QA Testing Not Applicable
- [x] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by ismaeldcom** [comment]
I've been thinking on the implementation and wonder if we could make HMR to work only for one entry. I think most of the time we will be working, and taking advantage of HMR, on only one entry. So we could try to enable the plugin only for that and get rid of the runtime. This will allow us to remove the PHP runtime load.

And we could then call it with something like: `npm run hmr -- <entry_name>`

What do you think?

File: includes/class-wc-payments.php [hunk 1]
```diff
@@ -1223,4 +1225,16 @@ public static function init_platform_checkout() {
...
+
+	/**
+	 * Load webpack runtime script, only if SCRIPT_DEBUG is enabled and the script exists.
+	 * Required for webpack server with HMR.
+	 *
+	 * @return void
+	 */
+	public static function enqueue_dev_runtime_scripts() {
+		if ( ( defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ) && file_exists( WCPAY_ABSPATH . 'dist/runtime.js' ) ) {
+			wp_enqueue_script( 'WCPAY_RUNTIME', plugins_url( 'dist/runtime.js', WCPAY_PLUGIN_FILE ), [], self::get_file_version( 'dist/runtime.js' ), true );
```
Comment (jrodger) [inline includes/class-wc-payments.php:1237]: I'm assuming `runtime.js` is created by HMR? My brief drive-by comment is to just check this new file can't get bundled up into the release ZIP accidentally. I think we destroy that `dist` folder before running the production build, so it's likely fine, but still worth double checking.

Comment (ismaeldcom) [reply includes/class-wc-payments.php:1237]: That's right, [I've also added](https://github.com/Automattic/woocommerce-payments/blob/5d19129845e132624bb08cf8f7bb325803f04516/webpack/shared.js#L35) `output.clean` webpack config, to ensure `runtime.js` doesn't get loaded while not using HMR (EG `npm run start`).

Anyway, not sure what do you think of my previous comment. I'd like to look at it and see if we can get rid of this conditional script load in the PHP side. Although, I'll need some time before getting into it 🙂

**Comment by fashuvo** [comment]
**Size Change:** -11 B (0%) 

**Total Size:** 1.69 MB

| Filename | Size | Change |
| :--- | :---: | :---: |
| `release/woocommerce-payments/dist/index.js` | 640 kB | -7 B (0%) |
| `release/woocommerce-payments/dist/onboarding.js` | 38.1 kB | -4 B (0%) |

<details><summary>ℹ️ <strong>View Unchanged</strong></summary>

| Filename | Size |
| :--- | :---: |
| `release/woocommerce-payments/assets/css/admin.css` | 913 B |
| `release/woocommerce-payments/assets/css/success.css` | 401 B |
| `release/woocommerce-payments/dist/blocks-checkout.css` | 1.77 kB |
| `release/woocommerce-payments/dist/blocks-checkout.js` | 28.1 kB |
| `release/woocommerce-payments/dist/checkout.css` | 1.85 kB |
| `release/woocommerce-payments/dist/checkout.js` | 26.7 kB |
| `release/woocommerce-payments/dist/index.css` | 35.8 kB |
| `release/woocommerce-payments/dist/multi-currency-analytics.js` | 8.41 kB |
| `release/woocommerce-payments/dist/multi-currency-switcher-block.js` | 79 kB |
| `release/woocommerce-payments/dist/multi-currency.css` | 14.7 kB |
| `release/woocommerce-payments/dist/multi-currency.js` | 117 kB |
| `release/woocommerce-payments/dist/onboarding.css` | 1.12 kB |
| `release/woocommerce-payments/dist/order.js` | 299 B |
| `release/woocommerce-payments/dist/payment-gateways.css` | 1.2 kB |
| `release/woocommerce-payments/dist/payment-gateways.js` | 91 kB |
| `release/woocommerce-payments/dist/payment-request.js` | 23 kB |
| `release/woocommerce-payments/dist/platform-checkout.css` | 3.73 kB |
| `release/woocommerce-payments/dist/platform-checkout.js` | 76.9 kB |
| `release/woocommerce-payments/dist/settings.css` | 47.2 kB |
| `release/woocommerce-payments/dist/settings.js` | 157 kB |
| `release/woocommerce-payments/dist/subscription-edit-page.js` | 3.66 kB |
| `release/woocommerce-payments/dist/subscription-product-onboarding-modal.css` | 7.18 kB |
| `release/woocommerce-payments/dist/subscription-product-onboarding-modal.js` | 16.7 kB |
| `release/woocommerce-payments/dist/subscription-product-onboarding-toast.js` | 428 B |
| `release/woocommerce-payments/dist/subscriptions-empty-state.css` | 298 B |
| `release/woocommerce-payments/dist/subscriptions-empty-state.js` | 49 kB |
| `release/woocommerce-payments/dist/tos.css` | 236 B |
| `release/woocommerce-payments/dist/tos.js` | 31.7 kB |
| `release/woocommerce-payments/dist/upe_checkout.css` | 1.85 kB |
| `release/woocommerce-payments/dist/upe_checkout.js` | 38 kB |
| `release/woocommerce-payments/dist/upe-blocks-checkout.css` | 1.77 kB |
| `release/woocommerce-payments/dist/upe-blocks-checkout.js` | 44.8 kB |
| `release/woocommerce-payments/includes/subscriptions/assets/css/plugin-page.css` | 633 B |
| `release/woocommerce-payments/includes/subscriptions/assets/js/plugin-page.js` | 355 B |
| `release/woocommerce-payments/vendor/automattic/jetpack-admin-ui/src/css/jetpack-icon.css` | 224 B |
| `release/woocommerce-payments/vendor/automattic/jetpack-assets/build/i18n-loader.js` | 2.43 kB |
| `release/woocommerce-payments/vendor/automattic/jetpack-assets/src/js/i18n-loader.js` | 1.01 kB |
| `release/woocommerce-payments/vendor/automattic/jetpack-identity-crisis/babel.config.js` | 160 B |
| `release/woocommerce-payments/vendor/automattic/jetpack-identity-crisis/build/index.css` | 2.27 kB |
| `release/woocommerce-payments/vendor/automattic/jetpack-identity-crisis/build/index.js` | 13.3 kB |
| `release/woocommerce-payments/vendor/automattic/jetpack-identity-crisis/build/index.rtl.css` | 2.27 kB |
| `release/woocommerce-payments/vendor/automattic/jetpack-tracking/src/js/tracks-ajax.js` | 789 B |
| `release/woocommerce-payments/vendor/automattic/jetpack-tracking/src/js/tracks-callables.js` | 925 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/about.css` | 1.2 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/admin-order-statuses.css` | 403 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/admin.css` | 3.43 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/checkout.css` | 299 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/modal.css` | 742 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/view-subscription.css` | 572 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/wcs-upgrade.css` | 411 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/admin-pointers.js` | 544 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/admin.js` | 8.97 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/jstz.js` | 6.8 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/jstz.min.js` | 3.83 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/meta-boxes-coupon.js` | 544 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/meta-boxes-subscription.js` | 2.35 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/moment.js` | 22.1 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/moment.min.js` | 11.6 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/payment-method-restrictions.js` | 1.29 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/wcs-meta-boxes-order.js` | 502 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/frontend/payment-methods.js` | 355 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/frontend/single-product.js` | 429 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/frontend/view-subscription.js` | 1.38 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/frontend/wcs-cart.js` | 387 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/modal.js` | 1.1 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/wcs-upgrade.js` | 1.27 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/build/index.css` | 392 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/build/index.js` | 3.06 kB |

</details>



<a href="https://github.com/preactjs/compressed-size-action"><sub>compressed-size-action</sub></a>

**Review by oaratovskyi — APPROVED** [review]
Thanks for wrangling this amazing feature! 😲 
Having components updated without refreshing the page is great. 💪 

I have checked the output of `build:release` command before/after - all good. Tried modifying text in the WCPay Setting page - works like a 🪄 

Let's 🚢  it!

**Comment by fashuvo** [comment]
**Size Change:** +5.99 kB (0%) 

**Total Size:** 1.69 MB

| Filename | Size | Change |  |
| :--- | :---: | :---: | :---: |
| `release/woocommerce-payments/dist/blocks-checkout.js` | 34 kB | +5.84 kB (+21%) | 🚨 |
| `release/woocommerce-payments/dist/checkout.js` | 26.9 kB | +157 B (+1%) |  |
| `release/woocommerce-payments/dist/index.js` | 640 kB | -7 B (0%) |  |
| `release/woocommerce-payments/dist/onboarding.js` | 38.1 kB | -4 B (0%) |  |

<details><summary>ℹ️ <strong>View Unchanged</strong></summary>

| Filename | Size |
| :--- | :---: |
| `release/woocommerce-payments/assets/css/admin.css` | 913 B |
| `release/woocommerce-payments/assets/css/success.css` | 401 B |
| `release/woocommerce-payments/dist/blocks-checkout.css` | 1.77 kB |
| `release/woocommerce-payments/dist/checkout.css` | 1.85 kB |
| `release/woocommerce-payments/dist/index.css` | 35.8 kB |
| `release/woocommerce-payments/dist/multi-currency-analytics.js` | 8.41 kB |
| `release/woocommerce-payments/dist/multi-currency-switcher-block.js` | 79 kB |
| `release/woocommerce-payments/dist/multi-currency.css` | 14.7 kB |
| `release/woocommerce-payments/dist/multi-currency.js` | 117 kB |
| `release/woocommerce-payments/dist/onboarding.css` | 1.12 kB |
| `release/woocommerce-payments/dist/order.js` | 299 B |
| `release/woocommerce-payments/dist/payment-gateways.css` | 1.2 kB |
| `release/woocommerce-payments/dist/payment-gateways.js` | 91 kB |
| `release/woocommerce-payments/dist/payment-request.js` | 23 kB |
| `release/woocommerce-payments/dist/platform-checkout.css` | 3.73 kB |
| `release/woocommerce-payments/dist/platform-checkout.js` | 76.9 kB |
| `release/woocommerce-payments/dist/settings.css` | 47.2 kB |
| `release/woocommerce-payments/dist/settings.js` | 157 kB |
| `release/woocommerce-payments/dist/subscription-edit-page.js` | 3.66 kB |
| `release/woocommerce-payments/dist/subscription-product-onboarding-modal.css` | 7.18 kB |
| `release/woocommerce-payments/dist/subscription-product-onboarding-modal.js` | 16.7 kB |
| `release/woocommerce-payments/dist/subscription-product-onboarding-toast.js` | 428 B |
| `release/woocommerce-payments/dist/subscriptions-empty-state.css` | 298 B |
| `release/woocommerce-payments/dist/subscriptions-empty-state.js` | 49 kB |
| `release/woocommerce-payments/dist/tos.css` | 236 B |
| `release/woocommerce-payments/dist/tos.js` | 31.7 kB |
| `release/woocommerce-payments/dist/upe_checkout.css` | 1.85 kB |
| `release/woocommerce-payments/dist/upe_checkout.js` | 38 kB |
| `release/woocommerce-payments/dist/upe-blocks-checkout.css` | 1.77 kB |
| `release/woocommerce-payments/dist/upe-blocks-checkout.js` | 44.8 kB |
| `release/woocommerce-payments/includes/subscriptions/assets/css/plugin-page.css` | 633 B |
| `release/woocommerce-payments/includes/subscriptions/assets/js/plugin-page.js` | 355 B |
| `release/woocommerce-payments/vendor/automattic/jetpack-admin-ui/src/css/jetpack-icon.css` | 224 B |
| `release/woocommerce-payments/vendor/automattic/jetpack-assets/build/i18n-loader.js` | 2.43 kB |
| `release/woocommerce-payments/vendor/automattic/jetpack-assets/src/js/i18n-loader.js` | 1.01 kB |
| `release/woocommerce-payments/vendor/automattic/jetpack-identity-crisis/babel.config.js` | 160 B |
| `release/woocommerce-payments/vendor/automattic/jetpack-identity-crisis/build/index.css` | 2.27 kB |
| `release/woocommerce-payments/vendor/automattic/jetpack-identity-crisis/build/index.js` | 13.3 kB |
| `release/woocommerce-payments/vendor/automattic/jetpack-identity-crisis/build/index.rtl.css` | 2.27 kB |
| `release/woocommerce-payments/vendor/automattic/jetpack-tracking/src/js/tracks-ajax.js` | 789 B |
| `release/woocommerce-payments/vendor/automattic/jetpack-tracking/src/js/tracks-callables.js` | 925 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/about.css` | 1.2 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/admin-order-statuses.css` | 403 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/admin.css` | 3.43 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/checkout.css` | 299 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/modal.css` | 742 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/view-subscription.css` | 572 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/css/wcs-upgrade.css` | 411 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/admin-pointers.js` | 544 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/admin.js` | 8.97 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/jstz.js` | 6.8 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/jstz.min.js` | 3.83 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/meta-boxes-coupon.js` | 544 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/meta-boxes-subscription.js` | 2.35 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/moment.js` | 22.1 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/moment.min.js` | 11.6 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/payment-method-restrictions.js` | 1.29 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/admin/wcs-meta-boxes-order.js` | 502 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/frontend/payment-methods.js` | 355 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/frontend/single-product.js` | 429 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/frontend/view-subscription.js` | 1.38 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/frontend/wcs-cart.js` | 387 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/modal.js` | 1.1 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/assets/js/wcs-upgrade.js` | 1.27 kB |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/build/index.css` | 392 B |
| `release/woocommerce-payments/vendor/woocommerce/subscriptions-core/build/index.js` | 3.06 kB |

</details>



<a href="https://github.com/preactjs/compressed-size-action"><sub>compressed-size-action</sub></a>
